File: week-1-c/section/contacts.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char *get_string(const char *prompt);
int *get_int(const char *prompt);

int main(void)
{
  char *name = get_string("What's your name? ");
  int *age = get_int("What's your age? ");
  char *phone_number = get_string("What's your phone number? ");

  printf("========================================\n");
  printf("Name: %s\n", name);
  printf("Age: %i\n", *age);
  printf("Phone Number: %s\n", phone_number);

  free(name);
  free(age);
  free(phone_number);

  return 0;
}

char *get_string(const char *prompt)
{
  int string_size = 255 * sizeof(char);
  printf("%s", prompt);

  char *str = malloc(string_size);
  if (str == NULL)
  {
    printf("Memory allocation failed.\n");
    exit(1);
  }

  fgets(str, string_size, stdin);
  str[strcspn(str, "\n")] = '\0';

  return str;
}

int *get_int(const char *prompt)
{
  int *i = malloc(sizeof(int));
  if (i == NULL)
  {
    printf("Memory allocation failed.\n");
    exit(1);
  }

  printf("%s", prompt);
  scanf("%i", i);

  int c;
  while ((c = getchar()) != '\n' && c != EOF)
  {
  }

  return i;
}

File: week-1-c/section/contacts/contacts.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	name, err := getString("What's your name? ")
	if err != nil {
		handleError(err)
		return
	}

	age, err := getInt("What's your age? ")
	if err != nil {
		handleError(err)
		return
	}

	phoneNumber, err := getString("What's your number? ")
	if err != nil {
		handleError(err)
		return
	}

	fmt.Println("========================================")
	fmt.Printf("Name: %s\n", name)
	fmt.Printf("Age: %d\n", age)
	fmt.Printf("Phone Number: %s\n", phoneNumber)
}

func getInput(prompt string) (string, error) {
	reader := bufio.NewReader(os.Stdin)

	fmt.Print(prompt)
	input, err := reader.ReadString('\n')
	if err != nil {
		return "", fmt.Errorf("error reading input: %v", err)
	}

	return strings.TrimSpace(input), nil
}

func getString(prompt string) (string, error) {
	return getInput(prompt)
}

func getInt(prompt string) (int, error) {
	for {
		input, err := getInput(prompt)
		if err != nil {
			return 0, err
		}

		num, err := strconv.Atoi(input)
		if err != nil {
			fmt.Println("invalid input. please enter a valid integer")
			continue
		}

		return num, nil
	}
}

func handleError(err error) {
	fmt.Fprintf(os.Stderr, "Error: %v\n", err)
}
